### PR TITLE
Fix FastifyInstance::ready type definition

### PR DESCRIPTION
**Pull request by jeromemacias**
#### Checklist

- [ ] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [x] commit message and code follows [Code of conduct](https://github.com/fastify/fastify/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by jeromemacias** [comment]
@delvedor done ;)

File: test/types/index.ts [hunk 1]
```diff
@@ -251,3 +251,18 @@ server.setSchemaCompiler(function (schema: object) {
...
+server.ready(function (err, done) {
+  if (err) throw err
+  done()
+})
+
+server.ready(function (err, context, done) {
+  if (err) throw err
+  server.log.debug(context)
+  done()
+})
```
Comment (mcollina) [inline test/types/index.ts]: You should do `done(err)`.

File: fastify.d.ts [hunk 2]
```diff
@@ -261,7 +261,9 @@ declare namespace fastify {
      * Registers a listener function that is invoked when all the plugins have
      * been loaded. It receives an error parameter if something went wrong.
      */
-    ready(readyListener?: () => void): void
+    ready(readyListener?: (err: Error) => void): void
+    ready(readyListener?: (err: Error, done: Function) => void): void
+    ready(readyListener?: (err: Error, context: FastifyInstance<HttpServer, HttpRequest, HttpResponse>, done: Function) => void): void
```
Comment (nwoltman) [inline fastify.d.ts]: If called without the `readyListener` parameter, a `Promise` that resolves to the `context` is returned.

**Comment by jeromemacias** [comment]
@mcollina @nwoltman Fixed and updated for promise case

File: fastify.d.ts [hunk 3]
```diff
@@ -261,7 +261,10 @@ declare namespace fastify {
      * Registers a listener function that is invoked when all the plugins have
      * been loaded. It receives an error parameter if something went wrong.
      */
-    ready(readyListener?: () => void): void
+    ready(): Promise<FastifyInstance<HttpServer, HttpRequest, HttpResponse>>
+    ready(readyListener?: (err: Error) => void): void
+    ready(readyListener?: (err: Error, done: Function) => void): void
+    ready(readyListener?: (err: Error, context: FastifyInstance<HttpServer, HttpRequest, HttpResponse>, done: Function) => void): void
```
Comment (nwoltman) [inline fastify.d.ts]: `readyListener` is not optional for the callback cases so the `?` shouldn't be needed.

Comment (jeromemacias) [reply fastify.d.ts]: You're right, good catch!

File: test/types/index.ts [hunk 4]
```diff
@@ -251,3 +251,26 @@ server.setSchemaCompiler(function (schema: object) {
 server.addSchema({})
 
 server.printRoutes()
+
+server.ready(function (err) {
+  if (err) throw err
+})
+
+server.ready(function (err, done) {
+  if (err) throw err
```
Comment (evanshortiss) [inline test/types/index.ts]: Should this if statement and throw be removed if using the callback?

File: test/types/index.ts [hunk 5]
```diff
@@ -251,3 +251,26 @@ server.setSchemaCompiler(function (schema: object) {
...
+  done(err)
+})
+
+server.ready(function (err, context, done) {
+  if (err) throw err
```
Comment (evanshortiss) [inline test/types/index.ts]: Same here?

File: test/types/index.ts (hunk 4 above)
Comment (delvedor) [reply test/types/index.ts]: There are not suggested paths on how handle this error, it's a decision that belongs to the user.
For the sake of the test, throw error is ok IMHO.

Comment (mcollina) [reply test/types/index.ts]: if there is a done callback, the error should be passed there.

Comment (mcollina) [reply test/types/index.ts]: and not thrown

**Pull request merged**